src/Evolution/Systems/RadiationTransport/M1Grey/M1Closure.cpp

```diff
@@ -0,0 +1,270 @@
+    const tnsr::i<DataVector, 3, Frame::Inertial>& momentum_density,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& fluid_velocity,
+    const Scalar<DataVector>& fluid_lorentz_factor,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>&
+        inv_spatial_metric) noexcept {
+  // Small number used to avoid divisions by zero
+  static const double TINY = 1.e-150;
+  // Below SMALL_VELOCITY, we use the v=0 closure,
+  // and we do not differentiate between fluid/inertial frames
+  static const double SMALL_VELOCITY = 1.e-15;
+  Variables<tmpl::list<LorentzFactorSquared, MomentumSquared, MomentumUp,
+                       VelocityLow, VelocitySquared>>
+      temp_closure_tensors(get(energy_density).size());
+
+  // The main calculation needed for the M1 closure is to find the
+  // roots of J^2 xi^2 = H^a H_a, with J the fluid-frame energy density
+  // and H^a the fluid-frame momentum density (0th and 1st moments).
+
+  // We first compute various fluid quantities and contractions
+  // with inertial moments needed even if v^2 is small
+  auto& w_sqr = get<LorentzFactorSquared>(temp_closure_tensors);
+  get(w_sqr) = square(get(fluid_lorentz_factor));
+  auto& v_sqr = get<VelocitySquared>(temp_closure_tensors);
+  get(v_sqr) = 1. - 1. / get(w_sqr);
+  // F^i
+  auto& f_up = get<MomentumUp>(temp_closure_tensors);
+  raise_or_lower_index(make_not_null(&f_up), momentum_density,
+                       inv_spatial_metric);
+  // F^i F_i
+  auto& f_sqr = get<MomentumSquared>(temp_closure_tensors);
+  get(f_sqr) = get<0>(momentum_density) * get<0>(f_up);
+  for (int m = 1; m < 3; m++) {
+    get(f_sqr) += momentum_density.get(m) * f_up.get(m);
+  }
+  // v_i
+  auto& v_low = get<VelocityLow>(temp_closure_tensors);
+  raise_or_lower_index(make_not_null(&v_low), fluid_velocity, spatial_metric);
+
+  // Loop over points
+  for (size_t s = 0; s < v_sqr.size(); ++s) {
+    const double& v2_pt = get(v_sqr)[s];
+    const double& w2_pt = get(w_sqr)[s];
+    const double& e_pt = get(energy_density)[s];
+    const double& f2_pt = (get(f_sqr)[s] > TINY ? get(f_sqr)[s] : TINY);
+    // Ignore complex closure calculations
+    // if the fluid velocity is very small
+    if (v2_pt < SMALL_VELOCITY) {
+      const double xi = sqrt(f2_pt) / e_pt;
+      const double chi = M1ClosureFunction(xi);
+      const double dthin = 1.5 * chi - 0.5;
+      const double dthick = 1. - dthin;
+
+      get(*closure_factor)[s] = xi;
+      get(*comoving_energy_density)[s] = e_pt;
+      get(*comoving_momentum_density_normal)[s] = 0.;
+      for (int i = 0; i < 3; i++) {
+        comoving_momentum_density_spatial->get(i)[s] =
+            momentum_density.get(i)[s];
+        // Closure assuming that fluid-frame = inertial frame
+        for (int j = i; j < 3; j++) {
+          pressure_tensor->get(i, j)[s] =
+              dthick * e_pt / 3. * inv_spatial_metric.get(i, j)[s] +
```

Comment:
rather than define dthin and dthick, define two variables one of which is `dthick*e_pt/3.`  and the other is `dthin*e_pt/f2_pt`